prov/util/src/util_cq.c

```diff
@@ -203,21 +203,38 @@ ssize_t ofi_cq_readerr(struct fid_cq *cq_fid, struct fi_cq_err_entry *buf,
 	struct util_cq *cq;
 	struct util_cq_err_entry *err;
 	struct slist_entry *entry;
-	ssize_t ret;
+	char *err_buf_save = NULL;
+	uint32_t api_version;
+	ssize_t ret = -FI_EAGAIN;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
+	api_version = cq->domain->fabric->fabric_fid.api_version;
+
 	fastlock_acquire(&cq->cq_lock);
-	if (!ofi_cirque_isempty(cq->cirq) &&
-	    (ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR)) {
-		ofi_cirque_discard(cq->cirq);
-		entry = slist_remove_head(&cq->err_list);
-		err = container_of(entry, struct util_cq_err_entry, list_entry);
-		*buf = err->err_entry;
-		free(err);
-		ret = 0;
-	} else {
-		ret = -FI_EAGAIN;
+	if (ofi_cirque_isempty(cq->cirq) ||
+		!(ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR))
+		goto unlock;
+
+	ofi_cirque_discard(cq->cirq);
+	entry = slist_remove_head(&cq->err_list);
+	err = container_of(entry, struct util_cq_err_entry, list_entry);
+	if ((FI_VERSION_GE(api_version, FI_VERSION(1, 5))) && buf->err_data_size) {
+		if (buf->err_data_size < err->err_entry.err_data_size) {
+			ret = -FI_ETOOSMALL;
```

Comment:
Btw, you rollback the list removal, but the cirque entry has still been discarded.